v0.17.1 panic when hub starting

#### What happened:

I0801 09:26:43.938440       1 flags.go:64] FLAG: --tls-private-key-file=""
I0801 09:26:43.938444       1 flags.go:64] FLAG: --tls-sni-cert-key="[]"
I0801 09:26:43.938451       1 flags.go:64] FLAG: --v="2"
I0801 09:26:43.938457       1 flags.go:64] FLAG: --version=""
I0801 09:26:43.938461       1 flags.go:64] FLAG: --vmodule=""
I0801 09:26:43.942590       1 hub.go:109] sending events to api server
I0801 09:26:43.942830       1 hub.go:145] starting clusternet-hub ...
I0801 09:26:44.454937       1 serving.go:374] Generated self-signed cert (apiserver.local.config/certificates/apiserver.crt, apiserver.local.config/certificates/apiserver.key)
I0801 09:26:44.456155       1 dynamic_serving_content.go:113] "Loaded a new cert/key pair" name="serving-cert::apiserver.local.config/certificates/apiserver.crt::apiserver.local.config/certificates/apiserver.key"
I0801 09:26:44.874020       1 requestheader_controller.go:244] Loaded a new request header values for RequestHeaderAuthRequestController
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2874ffd]

goroutine 1 [running]:
k8s.io/client-go/kubernetes.NewForConfig(0xc0003de640?)
        /go/pkg/mod/k8s.io/client-go@v0.29.0/kubernetes/clientset.go:461 +0x1d
github.com/clusternet/clusternet/pkg/hub/options.(*HubServerOptions).recommendedOptionsApplyTo(0xc0003b1d50, 0xc0006b4e00)
        /go/src/github.com/clusternet/clusternet/pkg/hub/options/options.go:247 +0xf0
github.com/clusternet/clusternet/pkg/hub/options.(*HubServerOptions).Config(0xc0003b1d50)
        /go/src/github.com/clusternet/clusternet/pkg/hub/options/options.go:161 +0x625
github.com/clusternet/clusternet/pkg/hub.(*Hub).Run(0xc000a837a0, {0x3f32a10, 0xc0002b9040})
        /go/src/github.com/clusternet/clusternet/pkg/hub/hub.go:146 +0xb4
github.com/clusternet/clusternet/cmd/clusternet-hub/app.NewClusternetHubCmd.func2(0xc0005db400?, {0xc0001b8c00?, 0x4?, 0x3ac1f37?})
        /go/src/github.com/clusternet/clusternet/cmd/clusternet-hub/app/app.go:91 +0x41d
github.com/spf13/cobra.(*Command).execute(0xc000004600, {0xc0000520a0, 0x8, 0x8})
        /go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:987 +0xaa3
github.com/spf13/cobra.(*Command).ExecuteC(0xc000004600)
        /go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1115 +0x3ff
github.com/spf13/cobra.(*Command).Execute(...)
        /go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1039
k8s.io/component-base/cli.run(0xc000004600)
        /go/pkg/mod/k8s.io/component-base@v0.29.0/cli/run.go:146 +0x2c5
k8s.io/component-base/cli.Run(0x3f32a10?)
        /go/pkg/mod/k8s.io/component-base@v0.29.0/cli/run.go:46 +0x17
main.main()
        /go/src/github.com/clusternet/clusternet/cmd/clusternet-hub/hub.go:34 +0xc5

#### What you expected to happen:

#### How to reproduce it (as minimally and precisely as possible):

#### Anything else we need to know?:

#### Environment:
- Clusternet version:
    - Clusternet-hub version: v0.17.1
- Kubernetes version (use `kubectl version`): 1.26
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Others:

--- a/pkg/hub/apiserver/shadow/apiserver.go
+++ b/pkg/hub/apiserver/shadow/apiserver.go
@@ -17,8 +17,9 @@ limitations under the License.
 package apiserver
 
 import (
+	"errors"
 	"fmt"
-	"path"
+	gpath "path"
 	"strings"
 	"time"
 
@@ -33,6 +34,7 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/managedfields"
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apiserver/pkg/admission"
 	genericapi "k8s.io/apiserver/pkg/endpoints"
@@ -48,6 +50,8 @@ import (
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/klog/v2"
 	apiservicelisters "k8s.io/kube-aggregator/pkg/client/listers/apiregistration/v1"
+	openapibuilder3 "k8s.io/kube-openapi/pkg/builder3"
+	openapicommon "k8s.io/kube-openapi/pkg/common"
 	custommetricsapi "k8s.io/metrics/pkg/apis/custom_metrics"
 	metricsapi "k8s.io/metrics/pkg/apis/metrics"
 
@@ -96,7 +100,13 @@ func init() {
 
 // ShadowAPIServer will make a shadow copy for all the APIs
 type ShadowAPIServer struct {
-	GenericAPIServer    *genericapiserver.GenericAPIServer
+	GenericAPIServer *genericapiserver.GenericAPIServer
+
+	// OpenAPIConfig will be used in generating OpenAPI spec.
+	OpenAPIConfig *openapicommon.Config
+	// OpenAPIV3Config will be used in generating OpenAPI V3 spec.
+	OpenAPIV3Config *openapicommon.OpenAPIV3Config
+
 	maxRequestBodyBytes int64
 	minRequestTimeout   int
 
@@ -123,7 +133,9 @@ func NewShadowAPIServer(apiserver *genericapiserver.GenericAPIServer,
 	kubeRESTClient restclient.Interface, clusternetclient *clusternet.Clientset,
 	manifestLister applisters.ManifestLister, apiserviceLister apiservicelisters.APIServiceLister,
 	crdInformerFactory crdinformers.SharedInformerFactory,
-	reservedNamespace string) *ShadowAPIServer {
+	reservedNamespace string,
+	openAPIConfig *openapicommon.Config, openAPIV3Config *openapicommon.OpenAPIV3Config,
+) *ShadowAPIServer {
 
 	return &ShadowAPIServer{
 		GenericAPIServer:    apiserver,
@@ -141,6 +153,8 @@ func NewShadowAPIServer(apiserver *genericapiserver.GenericAPIServer,
 			minRequestTimeout, maxRequestBodyBytes, admissionControl, apiserver.Authorizer, apiserver.Serializer, reservedNamespace),
 		apiserviceLister:  apiserviceLister,
 		reservedNamespace: reservedNamespace,
+		OpenAPIConfig:     openAPIConfig,
+		OpenAPIV3Config:   openAPIV3Config,
 	}
 }
 
@@ -251,15 +265,20 @@ func (ss *ShadowAPIServer) installAPIGroups(apiGroupInfos ...*genericapiserver.A
 		}
 	}
 
+	openAPIModels, err := ss.getOpenAPIModels(genericapiserver.APIGroupPrefix, apiGroupInfos...)
+	if err != nil {
+		return fmt.Errorf("unable to get openapi models: %v", err)
+	}
+
 	for _, apiGroupInfo := range apiGroupInfos {
-		if err := ss.installAPIResources(genericapiserver.APIGroupPrefix, apiGroupInfo); err != nil {
+		if err = ss.installAPIResources(genericapiserver.APIGroupPrefix, apiGroupInfo, openAPIModels); err != nil {
 			return fmt.Errorf("unable to install api resources: %v", err)
 		}
 
 		if apiGroupInfo.PrioritizedVersions[0].String() == shadowapi.SchemeGroupVersion.String() {
 			var found bool
 			for _, ws := range ss.GenericAPIServer.Handler.GoRestfulContainer.RegisteredWebServices() {
-				if ws.RootPath() == path.Join(genericapiserver.APIGroupPrefix, shadowapi.SchemeGroupVersion.String()) {
+				if ws.RootPath() == gpath.Join(genericapiserver.APIGroupPrefix, shadowapi.SchemeGroupVersion.String()) {
 					ss.crdHandler.SetRootWebService(ws)
 					found = true
 				}
@@ -300,7 +319,7 @@ func (ss *ShadowAPIServer) installAPIGroups(apiGroupInfos ...*genericapiserver.A
 
 // installAPIResources is a private method for installing the REST storage backing each api groupversionresource
 // copied from k8s.io/apiserver/pkg/server/genericapiserver.go and modified
-func (ss *ShadowAPIServer) installAPIResources(apiPrefix string, apiGroupInfo *genericapiserver.APIGroupInfo) error {
+func (ss *ShadowAPIServer) installAPIResources(apiPrefix string, apiGroupInfo *genericapiserver.APIGroupInfo, typeConverter managedfields.TypeConverter) error {
 	var resourceInfos []*storageversion.ResourceInfo
 	for _, groupVersion := range apiGroupInfo.PrioritizedVersions {
 		if len(apiGroupInfo.VersionedResourcesStorageMap[groupVersion.Version]) == 0 {
@@ -312,7 +331,7 @@ func (ss *ShadowAPIServer) installAPIResources(apiPrefix string, apiGroupInfo *g
 		if apiGroupInfo.OptionsExternalVersion != nil {
 			apiGroupVersion.OptionsExternalVersion = apiGroupInfo.OptionsExternalVersion
 		}
-
+		apiGroupVersion.TypeConverter = typeConverter
 		apiGroupVersion.MaxRequestBodyBytes = ss.maxRequestBodyBytes
 
 		discoveryAPIResources, r, err := apiGroupVersion.InstallREST(ss.GenericAPIServer.Handler.GoRestfulContainer)
@@ -394,4 +413,32 @@ func (ss *ShadowAPIServer) newAPIGroupVersion(apiGroupInfo *genericapiserver.API
 		Authorizer:          ss.GenericAPIServer.Authorizer,
 		MaxRequestBodyBytes: ss.maxRequestBodyBytes,
 	}
+}
+
+// getOpenAPIModels is a private method for getting the OpenAPI models
+// copied from k8s.io/apiserver/pkg/server/genericapiserver.go and modified
+func (ss *ShadowAPIServer) getOpenAPIModels(_ string, apiGroupInfos ...*genericapiserver.APIGroupInfo) (managedfields.TypeConverter, error) {
+	if ss.OpenAPIV3Config == nil {
+		// SSA is GA and requires OpenAPI config to be set
+		// to create models.
+		return nil, errors.New("OpenAPIV3 config must not be nil")
+	}
+	// hack for clusternet, no need for resourceNames
+	resourceNames := make([]string, 0)
+
+	// Build the openapi definitions for those resources and convert it to proto models
+	openAPISpec, err := openapibuilder3.BuildOpenAPIDefinitionsForResources(ss.OpenAPIV3Config, resourceNames...)
+	if err != nil {
+		return nil, err
+	}
+	for _, apiGroupInfo := range apiGroupInfos {
+		apiGroupInfo.StaticOpenAPISpec = openAPISpec
+	}
+
+	typeConverter, err := managedfields.NewTypeConverter(openAPISpec, false)
+	if err != nil {
+		return nil, err
+	}
+
+	return typeConverter, nil
 }

--- a/pkg/hub/hub.go
+++ b/pkg/hub/hub.go
@@ -213,7 +213,10 @@ func (hub *Hub) Run(ctx context.Context) error {
 					hub.clusternetInformerFactory.Apps().V1alpha1().Manifests().Lister(),
 					hub.aggregatorInformerFactory.Apiregistration().V1().APIServices().Lister(),
 					crdInformerFactory,
-					hub.options.ReservedNamespace)
+					hub.options.ReservedNamespace,
+					completeConfig.GenericConfig.OpenAPIConfig,
+					completeConfig.GenericConfig.OpenAPIV3Config,
+				)
 
 				crdInformerFactory.Start(postStartHookContext.StopCh)
 				return ss.InstallShadowAPIGroups(postStartHookContext.StopCh, hub.kubeClient.DiscoveryClient)

--- a/pkg/hub/options/options.go
+++ b/pkg/hub/options/options.go
@@ -19,18 +19,10 @@ package options
 import (
 	"fmt"
 	"net"
-	"net/http"
-	"strings"
-	"time"
 
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
-	"k8s.io/apimachinery/pkg/util/sets"
-	"k8s.io/apiserver/pkg/admission"
-	"k8s.io/apiserver/pkg/admission/plugin/namespace/lifecycle"
 	"k8s.io/apiserver/pkg/endpoints/openapi"
-	apirequest "k8s.io/apiserver/pkg/endpoints/request"
 	genericapiserver "k8s.io/apiserver/pkg/server"
-	genericfilters "k8s.io/apiserver/pkg/server/filters"
 	genericoptions "k8s.io/apiserver/pkg/server/options"
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	"k8s.io/client-go/dynamic"
@@ -39,7 +31,6 @@ import (
 	"k8s.io/client-go/rest"
 	cliflag "k8s.io/component-base/cli/flag"
 
-	clientset "github.com/clusternet/clusternet/pkg/generated/clientset/versioned"
 	informers "github.com/clusternet/clusternet/pkg/generated/informers/externalversions"
 	clusternetopenapi "github.com/clusternet/clusternet/pkg/generated/openapi"
 	"github.com/clusternet/clusternet/pkg/hub/apiserver"
@@ -127,41 +118,17 @@ func (o *HubServerOptions) Config() (*apiserver.Config, error) {
 		return nil, fmt.Errorf("error creating self-signed certificates: %v", err)
 	}
 
-	o.RecommendedOptions.ExtraAdmissionInitializers = func(c *genericapiserver.RecommendedConfig) ([]admission.PluginInitializer, error) {
-		client, err := clientset.NewForConfig(c.LoopbackClientConfig)
-		if err != nil {
-			return nil, err
-		}
-		informerFactory := informers.NewSharedInformerFactory(client, c.LoopbackClientConfig.Timeout)
-		o.LoopbackSharedInformerFactory = informerFactory
-		// TODO: add initializer
-		return []admission.PluginInitializer{}, nil
-	}
-
-	// remove NamespaceLifecycle admission plugin explicitly
-	o.RecommendedOptions.Admission.DisablePlugins = append(o.RecommendedOptions.Admission.DisablePlugins, lifecycle.PluginName)
-
 	serverConfig := genericapiserver.NewRecommendedConfig(apiserver.Codecs)
-	serverConfig.Config.RequestTimeout = time.Duration(40) * time.Second // override default 60s
-	serverConfig.LongRunningFunc = func(r *http.Request, requestInfo *apirequest.RequestInfo) bool {
-		if values := r.URL.Query()["watch"]; len(values) > 0 {
-			switch strings.ToLower(values[0]) {
-			case "true":
-				return true
-			default:
-				return false
-			}
-		}
-		return genericfilters.BasicLongRunningRequestCheck(sets.NewString("watch"), sets.NewString())(r, requestInfo)
-	}
-	serverConfig.OpenAPIConfig = genericapiserver.DefaultOpenAPIConfig(clusternetopenapi.GetOpenAPIDefinitions, openapi.NewDefinitionNamer(apiserver.Scheme))
-	serverConfig.OpenAPIConfig.Info.Title = openAPITitle
-	serverConfig.OpenAPIConfig.Info.Version = version.Get().GitVersion
-
 	if err := o.recommendedOptionsApplyTo(serverConfig); err != nil {
 		return nil, err
 	}
 
+	serverConfig.OpenAPIConfig = genericapiserver.DefaultOpenAPIConfig(clusternetopenapi.GetOpenAPIDefinitions, openapi.NewDefinitionNamer(apiserver.Scheme))
+	serverConfig.OpenAPIConfig.Info.Title = openAPITitle
+	serverConfig.OpenAPIConfig.Info.Version = version.Get().GitVersion
+	serverConfig.OpenAPIV3Config = genericapiserver.DefaultOpenAPIV3Config(clusternetopenapi.GetOpenAPIDefinitions, openapi.NewDefinitionNamer(apiserver.Scheme))
+	serverConfig.OpenAPIV3Config.Info.Title = openAPITitle
+	serverConfig.OpenAPIV3Config.Info.Version = version.Get().GitVersion
 	config := &apiserver.Config{
 		GenericConfig: serverConfig,
 		ExtraConfig:   apiserver.ExtraConfig{},
@@ -243,10 +210,21 @@ func (o *HubServerOptions) recommendedOptionsApplyTo(config *genericapiserver.Re
 	if err := o.RecommendedOptions.Audit.ApplyTo(&config.Config); err != nil {
 		return err
 	}
-
-	kubeClient, err2 := kubernetes.NewForConfig(config.ClientConfig)
-	if err2 != nil {
-		return err2
+	if err := o.RecommendedOptions.CoreAPI.ApplyTo(config); err != nil {
+		return err
+	}
+	var kubeClient *kubernetes.Clientset
+	var dynamicClient *dynamic.DynamicClient
+	if config.ClientConfig != nil {
+		var err error
+		kubeClient, err = kubernetes.NewForConfig(config.ClientConfig)
+		if err != nil {
+			return err
+		}
+		dynamicClient, err = dynamic.NewForConfig(config.ClientConfig)
+		if err != nil {
+			return err
+		}
 	}
 
 	if err := o.RecommendedOptions.Features.ApplyTo(&config.Config, kubeClient, config.SharedInformerFactory); err != nil {
@@ -257,10 +235,6 @@ func (o *HubServerOptions) recommendedOptionsApplyTo(config *genericapiserver.Re
 	}
 
 	initializers, err := o.RecommendedOptions.ExtraAdmissionInitializers(config)
-	if err != nil {
-		return err
-	}
-	dynamicClient, err := dynamic.NewForConfig(config.ClientConfig)
 	if err != nil {
 		return err
 	}

